core/src/main/scala/neotypes/Session.scala

```diff
@@ -17,21 +17,7 @@ final class Session[F[_]](private val session: NSession) extends AnyVal {
     }
 
   def transact[T](txF: Transaction[F] => F[T])(implicit F: Async[F]): F[T] =
-    transaction.flatMap { tx =>
-      val result = for {
-        r <- txF(tx)
-        _ <- tx.commit
-      } yield r
-
-      result.recoverWith {
-        case ex: Throwable =>
-          tx.rollback
-            .flatMap(_ => F.failed[T](ex))
-            .recoverWith {
-              case _ => F.failed[T](ex)
-            }
-      }
-    }
+    transaction.guarantee(txF)(_.commit)
```

Comment:
Not sure if I'm fully following the flow but it seems we always commit regardless of `txF` result. 
```scala
s.transact {tx =>
  for {
    res <- c"create (p:Record {personId: 1234)".query[Unit].execute(tx)
    _ <- Future.failure(new Exception("ERROR"))
  } yield ()
}
```

Reply:
Seems we'll need to add transaction tests like that one at some point

Reply:
@dimafeng Agree, I will add some tests to the transactions, just to make sure they work as intended for each effect.

Answering your first question, yes it will commit, regardless what happens.
My logic is that, on the scope of a `transact` there should only be logic related to **Neo4j** operations / queries. As such, the only possible errors, are those related with transactions themselves. And in those cases, the transaction will be automatically rollbacked before the commit, and at the end you will have the original exception _(the tests should confirm that!)_.

If that is not the case, and you have custom logic that may also fail. Then, `transact` is not the abstraction you should be using but `session.transaction.flatMap { tx => ??? }` and then a manual rollback or commit.

I am open to discussion if you believe the previous case is common enough... But, then the implementation becomes very tricky.

Reply:
I actually don't have strong opinion on this one. I'm open to the most expected (industry standard) behavior - I guess it's a question of what expected / industry standard behavior is 😁 

I think we can ask other people. cc @sloshy @gvolpe 

Maybe a stupid question, but I don't have answer off the top of my head - what would happen in case of:
```
s.transact {tx =>
  for {
    _ <- c"create (p:Record {personId: 1234)".query[Unit].execute(tx)
    _ <- c"broken cypher query".query[Unit].execute(tx)
  } yield ()
}
```

Reply:
Everything will be rollbacked by the **Neo4j java driver**. Thus, the moment we try to call `commit`, it will fail.

Again, the tests should confirm that, and that we preserve the original exception.
If that is not the case, then yes, we would need to change it.